Xenakios/TakeRenaming.cpp

```diff
@@ -223,40 +165,43 @@ void DoRenameTakeAndSourceFileDialog(COMMAND_T* ct)
 					if (newfnsplit[2].compare(fnsplit[2])!=0)
 						newfilename.append(fnsplit[2]);
 
-					MoveFile(oldname.c_str(),newfilename.c_str());
+					if (FileOrDirExists (newfilename.c_str()))
+					{
+						if (MessageBox(g_hwndParent, __LOCALIZE("Filename already exists!\nOverwrite?", "sws_DLG_165"), __LOCALIZE("Xenakios - Warning", "sws_DLG_165"), MB_OKCANCEL) == IDCANCEL)
+							continue;
+						else if (newfilename == oldname)
+								continue;
+					}
+					Main_OnCommand(40100, 0); // offline all media
+					MoveFile(oldname.c_str(), newfilename.c_str());
 					AddToRenameLog(oldname,newfilename);
 
-					for (int j=0;j<(int)alltakes.size();j++)
+					for (size_t j=0;j<alltakes.size();j++)
 					{
-						PCM_source *thesrc=(PCM_source*)GetSetMediaItemTakeInfo(alltakes[j],"P_SOURCE",0);
-						if (thesrc && thesrc->GetFileName() && strcmp(thesrc->GetType(),"SECTION")!=0)
-						{
-							string fname;
-							fname.assign(thesrc->GetFileName());
-							if (oldname.compare(fname)==0)
-							{
-								PCM_source *newsrc=PCM_Source_CreateFromFile(newfilename.c_str());
-								if (newsrc)
-								{
-									GetSetMediaItemTakeInfo(alltakes[j],"P_SOURCE",newsrc);
-									GetSetMediaItemTakeInfo(alltakes[j],"P_NAME",(char*)g_renameparams.NewName.c_str());
-									delete thesrc;
-
-									SNM_DeletePeakFile(oldname.c_str(), true); // no delete check (peaks files can be absent)
-								}
-							}
-						}
+						PCM_source* thesrcToRename = (PCM_source*)GetSetMediaItemTakeInfo(alltakes[j], "P_SOURCE", 0);
+						if (!thesrcToRename || strcmp(SWS_GetSourceFileName(thesrcToRename), oldname.c_str()))
+							continue;
+
+						thesrcToRename->SetFileName(newfilename.c_str());
+						// rename peak file, could be stored outside audiofiles folder (user pref)
+						char oldPeakFn[SNM_MAX_PATH] = ""; char newPeakFn[SNM_MAX_PATH] = "";
+						GetPeakFileName(oldname.c_str(),     oldPeakFn, sizeof(oldPeakFn));
+						GetPeakFileName(newfilename.c_str(), newPeakFn, sizeof(newPeakFn));
+						if(!MoveFile(oldPeakFn, newPeakFn))
+							SNM_DeletePeakFile(oldname.c_str(), true); // fallback if rename didn't work for some reason
```

Comment:
I found three situations where this breaks:

- The target directory (first two letters of the hashed filename) may not exist yet when storing peaks in the alternate path.
- `newPeakFn` is always in the alternate path even when the setting is disabled. The peak file is then moved to the wrong location (unless the above prevents it). This can be fixed by using `GetPeakFileNameEx` with forWrite=true.
- The move fails if across different filesystems (at least on non-Windows). This must be done via copy + delete.

```cpp
void SNM_MovePeakFile(const char *oldMediaFn, const char *newMediaFn)
{
	char oldPeakFn[SNM_MAX_PATH] {}, newPeakFn[SNM_MAX_PATH] {};
	GetPeakFileName(oldMediaFn, oldPeakFn, sizeof(oldPeakFn));
	GetPeakFileNameEx(newMediaFn, newPeakFn, sizeof(newPeakFn), true);

	if (const char dirSep { WDL_remove_filepart(newPeakFn) }) {
		CreateDirectory(newPeakFn, nullptr);
		newPeakFn[strlen(newPeakFn)] = dirSep;
	}

	if (!MoveFile(oldPeakFn, newPeakFn)) {
		SNM_CopyFile(newPeakFn, oldPeakFn);
		SNM_DeleteFile(oldPeakFn, true); // move to trash on Windows
	}
}
```

```suggestion
						SNM_MovePeakFile(oldname.c_str(), newfilename.c_str());
```